token-metadata/program/src/utils/mod.rs

```diff
@@ -190,18 +193,50 @@ pub fn zero_account(s: &str, size: usize) -> String {
 pub fn close_program_account<'a>(
     account_info: &AccountInfo<'a>,
     funds_dest_account_info: &AccountInfo<'a>,
+    key: Key,
 ) -> ProgramResult {
+    let rent = Rent::get()?;
+
+    // Metadata and Edition accounts could have fees stored, so we only want to withdraw
+    // the actual rent lamport amount. This will need to be further modified if fees are added
+    // to burn later.
+
+    let (rent_lamports, has_fees) = match key {
+        Key::MetadataV1 => (
+            rent.minimum_balance(Metadata::size()),
+            account_info.data.borrow()[METADATA_FEE_FLAG_INDEX] == FEE_FLAG_SET,
+        ),
+        _ => (account_info.lamports(), false),
+    };
+
+    let mut remaining_lamports = account_info
+        .lamports()
+        .checked_sub(rent_lamports)
+        .ok_or(MetadataError::NumericalOverflowError)?;
+
+    // If the account contains fees, remaining lamports will be the fee amount or the minimum rent exemption amount, whichever is greater.
+    if has_fees {
+        remaining_lamports = remaining_lamports.max(rent.minimum_balance(0));
+    }
+
+    let redeem_lamports = account_info
+        .lamports()
+        .checked_sub(remaining_lamports)
+        .ok_or(MetadataError::NumericalOverflowError)?;
+
     // Transfer lamports from the account to the destination account.
     let dest_starting_lamports = funds_dest_account_info.lamports();
-    **funds_dest_account_info.lamports.borrow_mut() = dest_starting_lamports
-        .checked_add(account_info.lamports())
-        .unwrap();
-    **account_info.lamports.borrow_mut() = 0;
+    **funds_dest_account_info.lamports.borrow_mut() =
+        dest_starting_lamports.checked_add(redeem_lamports).unwrap();
```

Comment:
Nit: for consistency this could also use `ok_or`

Reply:
good call